exporters/otlp/src/main/java/io/opentelemetry/exporter/otlp/AttributeMarshaler.java

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.exporter.otlp;
+
+import com.google.protobuf.ByteString;
+import com.google.protobuf.CodedOutputStream;
+import com.google.protobuf.WireFormat;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.proto.common.v1.AnyValue;
+import io.opentelemetry.proto.common.v1.ArrayValue;
+import io.opentelemetry.proto.common.v1.KeyValue;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+abstract class AttributeMarshaler extends MarshalerWithSize {
+  private static final AttributeMarshaler[] EMPTY_REPEATED = new AttributeMarshaler[0];
+  private final ByteString key;
+  private final int valueSize;
+
+  static AttributeMarshaler[] createRepeated(Attributes attributes) {
```

Comment:
It seems unfortunate to instantiate marshallers for each object. Isn't it better to define `Marshaler` something like

```java
interface MarshalerWithSize<T> {

  int serializedSize(T value);

  void writeValue(T value, CodedOutputStream stream);
}
```

and have singleton marshalers?

Reply:
Wilder would be to make everything static methods since we don't need to be generic, we only have a single API. But that's probably overboard.

Reply:
The problem is that the size of the key needs a conversion to utf8 which is very expensive and allocating one object I measured to be better and do the conversion once. We can always improve more once we have the initial structure in place

Reply:
I am looking also into a different approach that gogoproto uses which is to calculate size once and then write from the end to the buffer so to avoid the need of caching the size result per message. We have more ways to improve this and I will do a bit more on this during the holidays but still believe it is a good start

Reply:
Ok - also don't forget we have a fast concurrent weak map now, so it can often be used as an efficient cache when needed. If the attribute key is the main reason for the instantiation, then that'd be easy to cache and maybe a simple win.

Reply:
???